### PR TITLE
[Metrics] AUROC error on multilabel + improved testing

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Fixes https://github.com/PyTorchLightning/pytorch-lightning/issues/3303
Also improves metric testing by more tests against sklearn (related to #3230 )

# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by rohitgr7 — APPROVED** [review]

File: tests/metrics/functional/test_classification.py [hunk 1]
```diff
@@ -35,28 +38,43 @@
...
+    pytest.param(partial(sk_f1_score, average='macro'), f1_score, False, id='f1_score'),
+    pytest.param(partial(sk_fbeta_score, average='macro', beta=2),
+                 partial(fbeta_score, beta=2), False, id='fbeta_score'),
+    pytest.param(sk_confusion_matrix, confusion_matrix, False, id='confusion_matrix'),
+    pytest.param(sk_roc_curve, roc, True, id='roc'),
+    pytest.param(sk_precision_recall_curve, precision_recall_curve, True, id='precision_recall_curve'),
+    pytest.param(sk_roc_auc_score, auroc, True, id='auroc')
 ])
-def test_against_sklearn(sklearn_metric, torch_metric):
+def test_against_sklearn(sklearn_metric, torch_metric, binary):
```
Comment (Borda) [inline tests/metrics/functional/test_classification.py:53]: shall all be testes with binary?

Comment (SkafteNicki) [reply tests/metrics/functional/test_classification.py]: that is probably a good idea, I have changed the naming to `only_binary` such that when it is `False` both multiclass and binary is tested.

File: tests/metrics/functional/test_classification.py [hunk 2]
```diff
@@ -346,5 +364,13 @@ def test_iou(half_ones, reduction, remove_bg, expected):
     assert torch.allclose(iou_val, expected, atol=1e-9)
 
 
+@pytest.mark.parametrize('metric', [auroc])
+def test_error_on_multiclass_input(metric):
+    """ check that these metrics gives an error if they are used for multiclass problems  """
+    pred = torch.randint(0, 10, (100, ))
+    target = torch.randint(0, 10, (100, ))
+    with pytest.raises(ValueError):
```
Comment (awaelchli) [inline tests/metrics/functional/test_classification.py:372]:
```suggestion
    with pytest.raises(ValueError, match="AUROC metric is meant for binary classification"):
```

File: tests/metrics/functional/test_classification.py [hunk 3]
```diff
@@ -346,5 +364,13 @@ def test_iou(half_ones, reduction, remove_bg, expected):
     assert torch.allclose(iou_val, expected, atol=1e-9)
 
 
+@pytest.mark.parametrize('metric', [auroc])
+def test_error_on_multiclass_input(metric):
+    """ check that these metrics gives an error if they are used for multiclass problems  """
```
Comment (awaelchli) [inline tests/metrics/functional/test_classification.py:369]:
```suggestion
    """ check that these metrics raise an error if they are used for multiclass problems  """
```

File: tests/metrics/functional/test_classification.py [hunk 4]
```diff
@@ -35,28 +38,43 @@
...
+    pytest.param(partial(sk_f1_score, average='macro'), f1_score, False, id='f1_score'),
+    pytest.param(partial(sk_fbeta_score, average='macro', beta=2),
+                 partial(fbeta_score, beta=2), False, id='fbeta_score'),
+    pytest.param(sk_confusion_matrix, confusion_matrix, False, id='confusion_matrix'),
+    pytest.param(sk_roc_curve, roc, True, id='roc'),
+    pytest.param(sk_precision_recall_curve, precision_recall_curve, True, id='precision_recall_curve'),
+    pytest.param(sk_roc_auc_score, auroc, True, id='auroc')
 ])
-def test_against_sklearn(sklearn_metric, torch_metric):
+def test_against_sklearn(sklearn_metric, torch_metric, only_binary):
```
Comment (awaelchli) [inline tests/metrics/functional/test_classification.py:53]:
```suggestion
def test_against_sklearn(sklearn_metric, torch_metric, only_binary=False):
```
maybe this works (not sure), then you don't have to add it to all metrics above, only the few ones that are binary only

Comment (awaelchli) [reply tests/metrics/functional/test_classification.py:53]: the test has grown more complex now with the additional if statements. We have to be careful not to overload the tests with too much logic, or we have to test the test itself... if you know what I mean :)

**Review by awaelchli — APPROVED** [review]
LGTM, but be careful not to make tests too complicated.

File: tests/metrics/functional/test_classification.py (hunk 4 above)
Comment (awaelchli) [reply tests/metrics/functional/test_classification.py:53]: not a major concern, just something to look out in the future. I don't have a good suggestion as how to avoid it in this specific case.

**Comment by edenlightning** [comment]
mind looking at failing tests?

**Review by Borda — APPROVED** [review]
